lib/KVStore/KVStore.cpp

```diff
@@ -46,12 +53,59 @@ KVStore::KVImpl::KVImpl() {
     llvm::report_fatal_error((llvm::StringRef)"Redis connection error: " +
                              Ctx->errstr + "\n");
   }
+
+  if (!checkCompatibility()) {
+    llvm::report_fatal_error("Redis database on port %d is incompatible.", RedisPort);
+  }
 }
 
 KVStore::KVImpl::~KVImpl() {
   redisFree(Ctx);
 }
 
+bool KVStore::KVImpl::checkCompatibility() {
+  assert(Ctx && "Cannot check compatibility on an uninitialized database.");
+
+  redisReply *reply = static_cast<redisReply*>(redisCommand(Ctx, "GET cachetype"));
+  if (!reply || Ctx->err) {
+    llvm::report_fatal_error((llvm::StringRef)"Redis error: " + Ctx->errstr);
+  }
+
+  // get all current command line used
+  StringMap<cl::Option*> &Opts =  cl::getRegisteredOptions();
+  std::vector<std::string> ActiveOptions;
+  for (auto &K : Opts.keys()) {
+    if (Opts[K]->getNumOccurrences() && K.find("-debug-level") == StringRef::npos) {
+      ActiveOptions.emplace_back(K.str());
+    }
+  }
+  std::sort(ActiveOptions.begin(), ActiveOptions.end());
+  std::ostringstream ActiveOptionsOSS;
+  const char *delim = ",";
+  std::copy(ActiveOptions.begin(), ActiveOptions.end(),
+	    std::ostream_iterator<std::string>(ActiveOptionsOSS, delim));
+  std::string ActiveOptionsStr = ActiveOptionsOSS.str();
+
+  switch(reply->type) {
+  case REDIS_REPLY_NIL:
+    // no version set
+    freeReplyObject(reply);
+    reply = static_cast<redisReply*>(redisCommand(Ctx, "SET cachetype %s", ActiveOptionsStr.data()));
+    // TODO: Factor out all such snippets
+    if (!reply || Ctx->err) {
+      llvm::report_fatal_error((llvm::StringRef)"Redis error: " + Ctx->errstr);
+    }
+    break;
+  case REDIS_REPLY_STRING:
+    if (llvm::StringRef value = reply->str; value != ActiveOptionsStr)
+      return false;
+    break;
+  default: return false;
+  }
+
+  return true;
```

Comment:
this line is dead code, please remove

Reply:
this is not a dead code.

1) note that REDIS_REPLY_NIL and REDIS_REPLY_STRING don't return anything when things are compatible --- they just break out of the switch case and let this statement return true.
2) Even if I put return statements to return true as described in (1), compiler will emit warning saying that control reaches end of non-void statement as we don't disable `Wreturn-type` from our CMake config (which is fine).